pandas/core/reshape/reshape.py

```diff
@@ -1053,6 +1053,225 @@ def get_empty_frame(data) -> DataFrame:
+        if any(ps in col for ps in prefix_sep):
+            cat_columns.append(col)
+        else:
+            non_cat_columns.append(col)
+    data_to_decode = data[cat_columns].astype("boolean")
+    non_cat_data = data[non_cat_columns]
+
+    # get separator for each prefix and lists to slice data for each prefix
+    if isinstance(prefix_sep, dict):
+        variables_slice = {prefix: [] for prefix in prefix_sep}
+        for col in data_to_decode.columns:
+            for prefix in prefix_sep:
+                if prefix in col:
+                    variables_slice[prefix].append(col)
+    else:
+        sep_for_prefix = {}
+        variables_slice = {}
+        for col in data_to_decode.columns:
+            ps = [ps for ps in prefix_sep if ps in col][0]
+            prefix = col.split(ps)[0]
+            if prefix not in sep_for_prefix:
+                sep_for_prefix[prefix] = ps
+            if prefix not in variables_slice:
+                variables_slice[prefix] = [col]
+            else:
+                variables_slice[prefix].append(col)
+        prefix_sep = sep_for_prefix
+
+    # validate number of dropped_first
+    def check_len(item, name) -> None:
+        if not len(item) == len(variables_slice):
+            len_msg = (
+                f"Length of '{name}' ({len(item)}) did not match the "
+                "length of the columns being encoded "
+                f"({len(variables_slice)})."
+            )
+            raise ValueError(len_msg)
+
+    if dropped_first:
+        if isinstance(dropped_first, dict):
+            check_len(dropped_first, "dropped_first")
+        elif is_list_like(dropped_first):
+            check_len(dropped_first, "dropped_first")
+            dropped_first = dict(zip(variables_slice, dropped_first))
+        else:
+            dropped_first = dict(
+                zip(variables_slice, [dropped_first] * len(variables_slice))
+            )
+
+    cat_data = {}
+    for prefix, prefix_slice in variables_slice.items():
+        cats = [col[len(prefix + prefix_sep[prefix]) :] for col in prefix_slice]
+        assigned = data_to_decode[prefix_slice].sum(axis=1)
+        if any(assigned > 1):
+            raise ValueError(
+                f"Dummy DataFrame contains multi-assignment(s) for prefix: "
+                f"'{prefix}' in row {assigned.argmax()}."
+            )
+        elif any(assigned == 0):
+            if dropped_first:
+                cats.append(dropped_first[prefix])
+            else:
+                cats.append("from_dummies_nan_placeholer_string")
+            data_slice = concat((data_to_decode[prefix_slice], assigned == 0), axis=1)
+        else:
+            data_slice = data_to_decode[prefix_slice]
+        cat_data[prefix] = data_slice.dot(cats)
+
+    categorical_df = concat((non_cat_data, DataFrame(cat_data)), axis=1)
+    if dropped_first is None:
+        categorical_df.replace(
+            "from_dummies_nan_placeholer_string", np.nan, inplace=True
+        )
```

Comment:
I am still looking for a more elegant way of doing this

Reply:
yeah, sounds good